src/vanilla.ts

```diff
@@ -1,153 +1,179 @@
-    listener: StateListener<TState> | StateSliceListener<StateSlice>,
-    selector?: StateSelector<TState, StateSlice>,
-    equalityFn?: EqualityChecker<StateSlice>
-  ) => {
-    if (selector || equalityFn) {
-      return subscribeWithSelector(
-        listener as StateSliceListener<StateSlice>,
-        selector,
-        equalityFn
+type Create =
+  <T extends UnknownState, Mos extends [StoreMutatorIdentifier, unknown][] = []>
+    (initializer: StoreInitializer<T, [], Mos>) =>
+      Mutate<Store<T>, Mos>
+
+type StoreInitializer
+  < T extends UnknownState
+  , Mis extends [StoreMutatorIdentifier, unknown][]
+  , Mos extends [StoreMutatorIdentifier, unknown][]
+  , U = T
+  > =
+    & ( ( setState: A.Get<Mutate<Store<T>, Mis>, "setState", undefined>
+        , getState: A.Get<Mutate<Store<T>, Mis>, "getState", undefined>
+        , store: Mutate<Store<T>, Mis>
+        , $$storeMutations: Mis
+        ) =>
+          U
       )
-    }
-    listeners.add(listener as StateListener<TState>)
-    // Unsubscribe
-    return () => listeners.delete(listener as StateListener<TState>)
+    & { [$$storeMutators]?: Mos }
+declare const $$storeMutators: unique symbol;
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+interface StoreMutators<S, A> {}
+type StoreMutatorIdentifier = keyof StoreMutators<unknown, unknown>
+
+type Mutate<S, Ms> =
+  Ms extends [] ? S :
+  Ms extends [[infer Mi, infer Ma], ...infer Mrs]
+    ? Mutate<
+        StoreMutators<S, Ma>[Mi & StoreMutatorIdentifier],
+        Mrs
+      > :
+  never
+
+
+
+
+// ============================================================================
+// Implementation via Existential Types
+
+type EState = { __isState: true }
+type EShouldReplaceState = boolean & { __isShouldReplaceState: true }
+
+interface EStore
+  { getState:
+      () => EState
+  , setState:
+      ( nextStateOrUpdater:
+          | O.Partial<EState>
+          | ((state: EState) => O.Partial<EState>)
+      , shouldReplace?: EShouldReplaceState
+      ) =>
+        void
+  , subscribe:
+      ( listener:
+          (state: EState, previousState: E.Previous<EState>) => void
+      ) =>
+        () => void
+  , destroy:
+      () => void
+  }
+
+type ECreate =
+  ( storeInitializer:
+    ( set: EStore['setState']
+    , get: EStore['getState']
+    , store: EStore
+    ) => EState
+  ) =>
+    EStore
+
+const createImpl: ECreate = storeInitializer => {
+  let state: EState
+
+  type Listener = (state: EState, previousState: E.Previous<EState>) => void
+  const listeners: Set<Listener> = new Set()
+  const emit = (...a: F.Arguments<Listener>) =>
+    listeners.forEach(f => f(...a))
+
+  const store: EStore = {
+    getState: () => state,
+    setState: (nextStateOrUpdater, shouldReplace) => {
+      const nextState =
+        typeof nextStateOrUpdater === 'function'
+          ? nextStateOrUpdater(state)
+          : nextStateOrUpdater
+
+      if (objectIs(nextState, state)) return
```

Comment:
```suggestion
      if (nextState === state) return
```
We don't allow primitive values, so triple equals should just work? If so, we prefer keeping the current logic.

Reply:
There were some typing issues with `===` but now I rechecked and they're apparently gone so I'll revert to `===`.